internal/cache/file/downloader/parallel_downloads_job.go

```diff
@@ -58,61 +59,148 @@ func (job *Job) downloadRange(ctx context.Context, dstWriter io.Writer, start, e
-			if currGoRoutineIdx > 0 && !job.maxParallelismSem.TryAcquire(1) {
-				break
-			}
-
-			downloadErrGroup.Go(func() error {
-				if currGoRoutineIdx > 0 {
-					defer job.maxParallelismSem.Release(1)
-				}
-				// Copy the contents from NewReader to cache file at appropriate offset.
-				offsetWriter := io.NewOffsetWriter(cacheFile, rangeStart)
-				return job.downloadRange(downloadErrGroupCtx, offsetWriter, rangeStart, rangeEnd)
-			})
+	rangeMap := make(map[uint64]uint64)
+	job.offsetChan = make(chan offset, 2*job.fileCacheConfig.ParallelDownloadsPerFile)
+	var numGoRoutines int64
+	var start uint64
+	downloadChunkSize := uint64(job.fileCacheConfig.DownloadChunkSizeMb) * uint64(cacheutil.MiB)
+	downloadErrGroup, downloadErrGroupCtx := errgroup.WithContext(job.cancelCtx)
 
-			start = rangeEnd
+	// Start the goroutines as per the config and the availability.
+	for numGoRoutines = 0; (numGoRoutines < job.fileCacheConfig.ParallelDownloadsPerFile) && (start < job.object.Size); numGoRoutines++ {
+		// Respect max download parallelism only beyond first go routine.
+		if numGoRoutines > 0 && !job.maxParallelismSem.TryAcquire(1) {
+			break
 		}
 
-		// If any of the go routines failed, consider the async job failed.
-		err = downloadErrGroup.Wait()
-		if err != nil {
-			return
+		downloadErrGroup.Go(job.downloadOffsets(downloadErrGroupCtx, numGoRoutines, cacheFile, rangeMap))
+		start = start + downloadChunkSize
+	}
+
+	for start = 0; start < job.object.Size; {
+		nextOffset := offset{
+			start: start,
+			end:   min(job.object.Size, start+downloadChunkSize),
 		}
 
-		err = job.updateStatusOffset(start)
-		if err != nil {
-			return err
+		select {
+		case job.offsetChan <- nextOffset:
+			start = nextOffset.end
+			// In case we haven't started the goroutines as per the config, checking
+			// if any goroutines are available now.
+			// This may not be the ideal way, but since we don't have any way of
+			// listening if goroutines from other jobs have freed up, checking it here.
+			for numGoRoutines < job.fileCacheConfig.ParallelDownloadsPerFile && job.maxParallelismSem.TryAcquire(1) {
+				downloadErrGroup.Go(job.downloadOffsets(downloadErrGroupCtx, numGoRoutines, cacheFile, rangeMap))
+				numGoRoutines++
+			}
+		case <-downloadErrGroupCtx.Done():
+			return job.handleJobCompletion(downloadErrGroupCtx, start, downloadErrGroup)
 		}
 	}
-	return
+
+	return job.handleJobCompletion(downloadErrGroupCtx, start, downloadErrGroup)
+}
+
+// Job can be success or failure. This method will handle all the scenarios and
+// return the appropriate error.
+func (job *Job) handleJobCompletion(ctx context.Context, offset uint64, group *errgroup.Group) error {
+	// Close the channel since we are ending the job.
+	close(job.offsetChan)
+
+	// First check if the context has reported any error.
+	err := ctx.Err()
+	if err != nil {
+		// Also wait for all the goroutines to finish to ensure job is stopped.
+		waitErr := group.Wait()
```

Comment:
why can't we simply directly wait for `err = group.Wait` and then return that err instead of checking the ctx.Err separately ?

Reply:
updated the comment. In a scenario where context is cancelled before any goroutines are started or waiting for input. group.Wait doesn't return any error.

Ideally group.Wait is not required, but added to ensure we dont left any goroutines running.